### PR TITLE
More bitwise in Numbers.Natural.Abstract

### PR DESCRIPTION
**Pull request by Villetaneuse**
Added:
- `strong_induction_le` in NOrder (not sure about this one)
- `Even_Odd_induction` in NBits (not sure about capitalization)
- `land_(even|odd)_(even|odd)`, `land_le_[lr]`
- `ldiff_(even|odd)_(even|odd)`, `ldiff_le_l`
- `le_div2_diag_l`
- `le_shiftl` and `shiftr_le` (not sure about names here)
- `ones_0` and `ones_succ`

related to #17043: there, the lemmas are in a new file `NArith.Nbitwise` and proved with `lia`
@andres-erbsen `lnot_ones_same` is already there, named `lnot_ones`; I've not looked into `Positive` yet.

Questions:
- should this effort be more general (write similar lemmas for `lxor`, `lor`, ...)?
- What do you think of `Integer`? Should I try to give the most general statements in NatInt?

I don't think it's a final product, but I feel that I need some feedback now.

Fixes / closes #????

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.

- [ ] Opened **overlay** pull requests.

<!-- Pointers to relevant developer documentation:

Contributing guide: https://github.com/coq/coq/blob/master/CONTRIBUTING.md

Test-suite: https://github.com/coq/coq/blob/master/test-suite/README.md

Changelog: https://github.com/coq/coq/blob/master/doc/changelog/README.md

Building the doc: https://github.com/coq/coq/blob/master/doc/README.md
Sphinx: https://github.com/coq/coq/blob/master/doc/sphinx/README.rst
doc_gram: https://github.com/coq/coq/blob/master/doc/tools/docgram/README.md

Overlays: https://github.com/coq/coq/blob/master/dev/ci/user-overlays/README.md

### CODE REVIEW

**Comment by JasonGross** [comment]
Regarding high-level approach: I think the ultimate goal here should maybe be to provide a bitblast tactic that can (a) turn bitwise goals and hypotheses into equivalent boolean ones and then (b) discharge with btauto or similar.  And then have a library of "convenience lemmas" that are all proven with bitblast, and a core library of "specs of each operation in terms of testbit".

I'm not sure what the appropriate way to handle addition and multiplication would be---maybe turn every concrete number into a sum of powers of two, distribute multiplication over addition, turn multiplication by a power of two into left shift, have a rule that turns x + y into x | y when x & y = 0?  Or I guess we want (x + y) = x | y + x & y?

Thoughts?  (Apologies if this PR is the wrong place for this discussion and it belongs in a PR or CEP)

**Comment by andres-erbsen** [comment]
@JasonGross I think your high-level idea is different from the one I raised, but dependent on it: a bitwise-reasoning tactic would almost surely depend on Z through lia for side condiitons. Zbitwise.v has a tactic that kinda matches what you have in mind, I try not to call it bitblasting to avoid confusion with the easier known-width case.

**Comment by MSoegtropIMC** [comment]
I am also a fan of a split level approach: files with really basic stuff on top of which come tools like lia, bit blasting, you name it, on top of which come the user facing lemma library.

With the current structure a major problem ist that many useful lemmas, which could go into the library, are proven in a richer context, so that porting them to the standard library frequently requires a complete rewrite. This is bad.

**Comment by andres-erbsen** [comment]
I agree the considerations you bring up are relevant and sometimes in tension with using automation for ease of maintenance as I was advocating; we'll have to see for each proposed addition. As for the user-visible dependency graph, my preference would be for NArith to be the path for all theory about N operations and that we should make a new path for only basic theory for (probably less common) use cases that require it.

@coqbot run full ci

**Comment by andres-erbsen** [comment]
CI looks good; windows failure looks unrelated. I believe the precedent is to have a changelog entry for this. I intend to look at this again in a couple of days and to merge if it has a changelog entry by then.

**Comment by Villetaneuse** [comment]
Thank you!
If you don't mind, I would also like to:
- [ ] change `binary_induction` (and redefine it without the spurious `Proper` clause in implementations) so that it can be useful with `induction with` : this could be difficult (at least
- [ ] double check the proofs to add missing non-controversial lemmas
- [ ] double check the Search outputs in PeanoNat and BinNat to make sure there is no obvious overlapping

If it's ok and people use these lemmas we can easily add more for `lor`, `lxor`, etc; but this can be done in a separate PR.

File: theories/Numbers/Natural/Abstract/NBits.v
Comment (proux01) [inline theories/Numbers/Natural/Abstract/NBits.v:600]: Maybe something like `0 < a -> div2 a < a` could also be useful sometimes?

**Comment by proux01** [comment]
> I would be willing to assign (and merge after working out minor details) but I'll give other stdlib devs the opportunity since this design is substantially influenced by my own work and independent review is good or something.

I only had a quick look and this mostly looks reasonable. You two are certainly more knowledgable than me about that part of the library, so please go ahead.

> the Numbers-based proofs here are 5x longer than the lia-based proofs versions [here](https://github.com/coq/coq/pull/17043/files#diff-fdeb6e2b99db68716e207edb45f911381576f9c9d317b228819ab3b1bc3350eb), and I'd guess they took more than 5x as long to create (thanks again @Villetaneuse!). I think this comparison is a clear indication that we would benefit from separate `Require`-able modules for as little and as much of NArith as possible, perhaps NArithMinimal (NArithBase?) and NArithMaximal (NArith?). Porting lia-based proofs to the current structure is needless friction for development and maintenance. (I don't think we should block this PR on it, but I would like to find a way forward here)

It seems we all agree. A PR clearly separating some NArithBase from some more general files that could use lia would certainly be welcome. It could for instance put the current content of `theories/NArith` (or a subset of it) in some `theories/NArith/Base` directory and we would then have a `theories/NArith/All` directory whose content would require lia. Ideally a CI job would check that NArith/Base can be compiled without more dependencies (but I could do that later, for instance as part of the package split while implementing the stdlib CEP https://github.com/coq/ceps/pull/83 ).

File: theories/Numbers/Natural/Abstract/NBits.v
Comment (Villetaneuse) [reply theories/Numbers/Natural/Abstract/NBits.v:600]: ok, but I prefer `a <> 0` instead of `0 < a`, since it is solved by discriminate.

Comment (proux01) [reply theories/Numbers/Natural/Abstract/NBits.v:600]: Sure, I'm spoiled by `ssrnat.leq` that computes ;) (making even discriminate useless here).

**Comment by Villetaneuse** [comment]
Sorry for the noise... I made many mistakes I will fix tomorrow. My poor brain still struggles with git.

**Comment by Villetaneuse** [comment]
No real progress (except a crazy amount of git shenanigans to fix whitespace issues and errors on my part).
There are lemmas in `PeanoNat` which already cover some of these.
- `le_div2` \approx `le_div2_diag_l`
- `lt_div2` \approx 'lt_div2_diag_l`
- `div2_double` = `div2_even`
- `div2_odd'` \approx `div2_succ_double`

I need to work this out. The best solution would be to move these lemmas from PeanoNat to Numbers/Natural so that we get them at least in N and Nat (maybe also in Z, but I'm scared by division in Z).

The naming of `odd_succ_double`, etc is still an issue.

I still can't figure how I could remove the "Proper" condition in `PeanoNat`. I'm considering killing `strong_induction_le ` and making `binary_induction` Private.

**Comment by JasonGross** [comment]
> I still can't figure how I could remove the "Proper" condition in `PeanoNat`. I'm considering killing `strong_induction_le ` and making `binary_induction` Private

If you're talking about `Proper (eq ==> iff) A`, you don't get to remove this until you reach the level of concreteness that tells you that `eq` is leibniz equality.  You could also remove it by phrasing the conclusion as `forall n, exists m, n == m /\ A m`.  As a counterexample, consider trying to instantiate this module with `{ q : Q | q == Qround q }` (the rationals representing whole numbers).  If `A` demands that the rational is also in reduced form, then you can't just use induction to prove the property.

**Comment by Villetaneuse** [comment]
> > I still can't figure how I could remove the "Proper" condition in `PeanoNat`. I'm considering killing `strong_induction_le ` and making `binary_induction` Private
> 
> If you're talking about `Proper (eq ==> iff) A`, you don't get to remove this until you reach the level of concreteness that tells you that `eq` is leibniz equality. You could also remove it by phrasing the conclusion as `forall n, exists m, n == m /\ A m`. As a counterexample, consider trying to instantiate this module with `{ q : Q | q == Qround q }` (the rationals representing whole numbers). If `A` demands that the rational is also in reduced form, then you can't just use induction to prove the property.

Thank you for your comment! I was not saying that the `Proper` conditions are useless in general. There are indeed many models of this theory where equality isn't Leibniz. But I'm convinced that in this case where the actual used implementations of natural numbers and integers all use syntactic equality, it was a mistake to aim for this level of generality. All the user gets is an environment full of spurious conditions; the biggest one being `Proper eq ==> A` but there are many others; many lemmas about natural numbers have some `0 <= n` condition; which is embarrassing.

I certainly don't want to add another one of these suboptimal lemmas, hence my problem with my `binary_induction`.

**Comment by andres-erbsen** [comment]
One possibility for specializing the lemmas is to do so after the Numbers abstraction is instantiated with concrete Inductives. The general lemmas could even be named Private_ so they don't show up in Search.

I also think the setoid equality is a bad choice, but it is still used in https://github.com/coq-community/bignums/blob/3885b1a3ff6475c692fc83b539a43b8558fd1ec8/BigN/NMake.v#L34 so I don't think there is a trivial way to get rid of it.

**Comment by Villetaneuse** [comment]
> One possibility for specializing the lemmas is to do so after the Numbers abstraction is instantiated with concrete Inductives. The general lemmas could even be named Private_ so they don't show up in Search.

Do you mean that I should call `Private_` the weak induction lemmas and then use this `Private_` version to prove a stronger (without `Proper`) lemma in `PeanoNat` and `BinNat`?

> 
> I also think the setoid equality is a bad choice, but it is still used in https://github.com/coq-community/bignums/blob/3885b1a3ff6475c692fc83b539a43b8558fd1ec8/BigN/NMake.v#L34 so I don't think there is a trivial way to get rid of it.

No it would clearly break too many things... these are certainly the most used parts of the stdlib. One possibility (in further PR's) would be to deprecate (or just hide) the weaker lemmas and provide stronger alternatives.

**Comment by Villetaneuse** [comment]
I think I'm mostly done with this PR, unless of course you have suggestions.
In the end:
- I moved a part of `PeanoNat` in `NBits`, it's better for consistency between `N` and `nat`. Some lemmas could have better names (in particular `div2_decr`) but I'm not sure I want to break libraries over this.
- no solution for `div2_double`: it's two different things in `PeanoNat` and `BinInt`; I chose to keep `div2_even` (even though it's a duplicate of `div2_double` there) and `div2_odd'`.
- my new `(even|odd)_[succ_]double` are renamed `(even|odd)_(even|odd)`.
- This way, the naming scheme is clearer: `even` means `(2 * n)` and `odd` means `2 * n + 1`; it's probably not perfect but it's consistent with previous names about `testbit`.
- There is a test file to check that the "induction" lemmas work with the `induction` tactic.

@coq/stdlib-maintainers @coq/number-maintainers please feel free to criticize this work, changing both `PeanoNat` and `BinNat` should not be taken lightly.

When/if it's fine for you I will write a detailed changelog entry.

File: theories/Arith/PeanoNat.v
Comment (andres-erbsen) [inline theories/Arith/PeanoNat.v:637]: maybe `unfold_double`, or flip the sides and `mul_2_l`?

File: theories/Arith/PeanoNat.v
Comment (andres-erbsen) [inline theories/Arith/PeanoNat.v:688]: @coq/number-maintainers can you confirm this approach is reasonable?

File: theories/Numbers/NatInt/NZParity.v
Comment (andres-erbsen) [inline theories/Numbers/NatInt/NZParity.v:269]: What do you think about perhaps using double and succdouble in these and related lemmas, either in just the names or in names and definitions?

File: theories/Numbers/Natural/Abstract/NBits.v
Comment (andres-erbsen) [inline theories/Numbers/Natural/Abstract/NBits.v:816]: ...for example, `lande_double_l`?

File: theories/Numbers/Natural/Abstract/NMulOrder.v
Comment (andres-erbsen) [inline theories/Numbers/Natural/Abstract/NMulOrder.v:59]: I initially read these names as indicating which side of `le` the `mul` is on... I don't have any better ideas

File: theories/Numbers/Natural/Abstract/NPow.v
Comment (andres-erbsen) [inline theories/Numbers/Natural/Abstract/NPow.v:125]: `le_1_pow`?

File: theories/Numbers/Natural/Abstract/NSub.v
Comment (andres-erbsen) [inline theories/Numbers/Natural/Abstract/NSub.v:254]: `sub_sub_r`?

File: theories/Arith/PeanoNat.v
Comment (andres-erbsen) [inline theories/Arith/PeanoNat.v:924]: Should there be a version of this lemma for N?

**Review by andres-erbsen — APPROVED** [review]

**Comment by andres-erbsen** [comment]
> This way, the naming scheme is clearer: even means (2 * n) and odd means 2 * n + 1; it's probably not perfect but it's consistent with previous names about testbit.

The concern I have with this is that there is already a predicate called `even`. I can't make sense why the previous lemmas were named that way. I don't have strong feelings, but unless you do, I'd ignore the inexplicable examples or even add aliases for them with new names.

**Comment by Villetaneuse** [comment]
> > This way, the naming scheme is clearer: even means (2 * n) and odd means 2 * n + 1; it's probably not perfect but it's consistent with previous names about testbit.
> 
> The concern I have with this is that there is already a predicate called `even`. I can't make sense why the previous lemmas were named that way. I don't have strong feelings, but unless you do, I'd ignore the inexplicable examples or even add aliases for them with new names.

I share your concerns. I didn't find an easy solution. `double` and  `succ_double` are already taken (and mean different things in `nat` and `N`), `mul_2_l` could be ok for even, but then `add_1_r_mul_2_l` is horrible, ...

My line of thinking was that it's ok as long as it's not in front, `even_...` will be lemmas stating stuff about the `even` boolean predicate and `op_even` is about `op (2 * n)`. I know it's not perfect at all, let me think again about it a little bit.

**Comment by andres-erbsen** [comment]
@Villetaneuse what do you think would be the best path forward here? I don't feel strongly enough about any of my suggestions to hold this back even further if you think we should just merge.

**Comment by andres-erbsen** [comment]
I intend to merge this PR in a week or so -- please speak up if you object or if you would like small change to go in with it.

File: theories/Arith/PeanoNat.v
Comment (proux01) [inline theories/Arith/PeanoNat.v:909]: Wouldn't
```suggestion
Export PrivateImplementsBitwiseSpec.
```
be enough here?

**Comment by proux01** [comment]
This lacks a changelog entry but otherwise looks reasonnable.

**Comment by andres-erbsen** [comment]
I have drafted a changelog here https://github.com/andres-erbsen/coq/tree/more_bitwise_in_Natural, with a small fixup in the process, but I do not have permissions to push to the PR branch and I believe the author is busy with other stuff. Should I just create another PR?

**Comment by proux01** [comment]
@andres-erbsen strange, the "Maintainers are allowed to edit this pull request." box is checked and I was able to push your commit.

**Comment by andres-erbsen** [comment]
I think for Github (as opposed to Coqbot), maintainers means core devs.

**Comment by proux01** [comment]
That sounds bound, it shouldn't IMO.
Let's ping @Zimmi48 who may know better: can't maintainers push to PRs with "Maintainers are allowed to edit this pull request"?

**Comment by JasonGross** [comment]
At a glance, https://github.com/orgs/coq/teams/contributors?query=role%3Amaintainer+membership%3Achild-team and https://github.com/orgs/coq/teams/contributors?query=role%3Amaintainer seem to contain only the core devs indeed.

**Comment by proux01** [comment]
But https://github.com/orgs/coq/teams/contributors?query=role%3Amember+membership%3Achild-team seems to contain everybody. I don't understand much to those github things, let's add it to the agenda of a forthcoming call: https://github.com/coq/coq/wiki/Coq-Call-2024-04-30

**Comment by JasonGross** [comment]
> But https://github.com/orgs/coq/teams/contributors?query=role%3Amember+membership%3Achild-team seems to contain everybody

Yes, that is everybody with write access I believe.  There's a table of permissions [here](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization).  Given that it specifically says "maintainers" in  "Maintainers are allowed to edit this pull request", it's not surprising that people with only write access but not maintain access can't push.

**Comment by andres-erbsen** [comment]
I do see that text :thinking:.

**Comment by JasonGross** [comment]
Did you add git@github.com:Villetaneuse/coq.git as your remote or https://github.com/Villetaneuse/coq.git, or did you try to push to refs/pull/18628/head on coq/coq?  I believe only the first option (git@github.com:Villetaneuse/coq.git) will work.

**Comment by Villetaneuse** [comment]
@coqbot run full ci

**Comment by andres-erbsen** [comment]
@coqbot merge now

**Pull request merged**
